src/transformers/integrations/peft.py

```diff
@@ -159,12 +161,63 @@ def load_adapter(
                 used for inference.
+            hotswap : (`"auto"` or `bool`, *optional*, defaults to `"auto"`)
+                Whether to substitute an existing (LoRA) adapter with the newly loaded adapter in-place. This means
+                that, instead of loading an additional adapter, this will take the existing adapter weights and replace
+                them with the weights of the new adapter. This can be faster and more memory efficient. However, the
+                main advantage of hotswapping is that when the model is compiled with torch.compile, loading the new
+                adapter does not require recompilation of the model. When using hotswapping, the passed `adapter_name`
+                should be the name of an already loaded adapter.
+
+                If the new adapter and the old adapter have different ranks and/or LoRA alphas (i.e. scaling), you need
+                to call an additional method before loading the adapter:
+
+                ```py
+                model = AutoModel.from_pretrained(...)
+                max_rank = ...  # the highest rank among all LoRAs that you want to load
+                # call *before* compiling and loading the LoRA adapter
+                model.enable_peft_hotswap(target_rank=max_rank)
+                model.load_adapter(file_name_1, adapter_name="default")
+                # optionally compile the model now
+                model = torch.compile(model, ...)
+                output_1 = model(...)
+                # now you can hotswap the 2nd adapter, use the same name as for the 1st
+                # hotswap is activated by default since enable_peft_hotswap was called
+                model.load_adapter(file_name_2, adapter_name="default")
+                output_2 = model(...)
+                ```
+
+                By default, hotswap is disabled and requires passing `hotswap=True`. If you called
+                `enable_peft_hotswap` first, it is enabled. You can still manually disable it in that case by passing
+                `hotswap=False`.
+
+                Note that hotswapping comes with a couple of limitations documented here:
+                https://huggingface.co/docs/peft/main/en/package_reference/hotswap
             adapter_kwargs (`dict[str, Any]`, *optional*):
                 Additional keyword arguments passed along to the `from_pretrained` method of the adapter config and
                 `find_adapter_config_file` method.
         """
+        from peft import PeftType
+
         check_peft_version(min_version=MIN_PEFT_VERSION)
 
+        if hotswap == "auto":
+            # if user called model.enable_peft_hotswap and this is not the first adapter, enable hotswap
+            hotswap_enabled = getattr(self, "_hotswap_enabled", False)
+            not_first_adapter = bool(self._hf_peft_config_loaded and (adapter_name in self.peft_config))
```

Comment:
We should not need bool casting here, do we?
```suggestion
            not_first_adapter = self._hf_peft_config_loaded and adapter_name in self.peft_config
```

Reply:
Right now, `_hf_peft_config_loaded` is a bool, so coercing is not needed. But if its type changes in the future, the type of `not_first_adapter` could also change, so the cast makes this line future proof. Just an example:

```python
>>> _hf_peft_config_loaded = {"foo": 1}
>>> adapter_name = "default"
>>> peft_config = {"default": 2}
>>> not_first_adapter = _hf_peft_config_loaded and adapter_name in peft_config
>>> not_first_adapter
True
>>> 
>>> _hf_peft_config_loaded = {}
>>> not_first_adapter = _hf_peft_config_loaded and adapter_name in peft_config
>>> not_first_adapter
{}
```

As you can see, in the last line, we suddenly have a different type for `not_first_adapter`.